### PR TITLE
add preprocess 

### PR DESCRIPTION
**Pull request by matt-long**
This PR enables a `preprocess` argument to `to_dataset_dict`. For example:
```python
    def rename_coords(ds):
        return ds.rename({'lon': 'longitude', 'lat': 'latitude'})
    
    col = intake.open_esm_datastore(esmcol_path)
    col_sub = col.search(**query)
    
    dsets = col_sub.to_dataset_dict(preprocess=rename_coords)
```
cc @rabernat @jbusecke

### CODE REVIEW

**Comment by andersy005** [comment]
Towards #152

**Comment by jbusecke** [comment]
This functionality does not work as expected for me: 

Quick example:
```python
import intake
query = dict(experiment_id='historical', table_id='Oyr', 
                 variable_id='o2', grid_label='gn', member_id='r1i1p1f1', source_id='CESM2-WACCM')
cat = col.search(**query)

def test_func(ds):
    if 'nlon' in ds.dims:
        ds = ds.rename({'nlon':'WAZOOOOOONGA'})
    return ds

dsets_pp = cat.to_dataset_dict(preprocess=test_func)

dsets_pp['CMIP.NCAR.CESM2-WACCM.historical.Oyr.gn'].dims
```
gives `Frozen(SortedKeysDict({'nlat': 384, 'nlon': 320, 'vertices': 4, 'time': 165, 'd2': 2, 'lev': 60, 'member_id': 1}))
`
When I apply the function straight to the dataset I get this:
```
test_func(dsets_pp['CMIP.NCAR.CESM2-WACCM.historical.Oyr.gn']).dims
```

I get the expected change `Frozen(SortedKeysDict({'d2': 2, 'lev': 60, 'member_id': 1, 'nlat': 384, 'WAZOOOOOONGA': 320, 'time': 165, 'vertices': 4}))
`

Am I missing something or is this a bug?

**Comment by matt-long** [comment]
@jbusecke, we are not able to reproduce this error. Are you using the most recent collection spec from 
https://github.com/NCAR/intake-esm-datastore/tree/master/catalogs
?

**Comment by jbusecke** [comment]
No. Let me try again in a minute.

**Comment by jbusecke** [comment]
I updated to the latest and it still produces the same error. See here.
https://github.com/jbusecke/cmip6_preprocessing/blob/master/notebooks/testing_intake_esm_preprocessin.ipynb

**Comment by andersy005** [comment]
> Is the version not the latest?

Yep

**Comment by jbusecke** [comment]
I updated the notebook to double check that there is no error on my side. Could you pull it and try to run it on your side?

https://nbviewer.jupyter.org/github/jbusecke/cmip6_preprocessing/blob/master/notebooks/testing_intake_esm_preprocessin.ipynb

**Comment by matt-long** [comment]
Ok, I think we have this pinned down.

You are calling  `to_dataset_dict` twice, first without the `preprocess` argument. On the second call when you pass this in, however, `intake-xarray` (under the hood) is simply returning the cached dataset on the object: it is not detecting the change in the input args to appropriately update output. 

I am fixing this behavior now.

**Comment by rabernat** [comment]
It would be ideal if `to_dataset_dict` should not have [side effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science)).

**Comment by jbusecke** [comment]
Thanks for the quick fix @andersy005, I am now able to use custom preprocessing functions! 

I made a quick example [here](https://github.com/jbusecke/cmip6_preprocessing/blob/master/notebooks/testing_intake_esm_preprocessin.ipynb), should we include that in the docs/readme?

**Comment by naomi-henderson** [comment]
@jbusecke and @andersy005 , this is a great example for the documentation! 

By the way,  which `col' are you using, cloud or glade?  It doesn't seem to be specified - but you do call `to_dataset_dict` without adding the `zarr_kwargs` or `cdf_kwargs`.  If it is `pangeo-cmip6.json`, not using `zarr_kwargs={'consolidated': True}` is very slow.

Since your example uses your `cmip6_preprocessing` package, it would be good to point to your repo or conda/pip install method.  How about using a simple pre-processing function defined in the notebook itself, instead?

**Comment by jbusecke** [comment]
See #168
